{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e46a429c-b4dd-416f-bc87-98874700693d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd \n",
    "import argparse\n",
    "import numpy as np \n",
    "import glob\n",
    "import time \n",
    "import logging\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0b8ee29d-8559-4779-9c3e-750816518ffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_bed(bed, outname, p5nt = ['G', 'A'], min_len = 21, max_len = 23, orientation = '-') : \n",
    "    \n",
    "    output = open(outname, 'w')\n",
    "    lines = ''\n",
    "    with open(bed, 'r') as f : \n",
    "        for line in f :\n",
    "            if not line.startswith(\"gene\") : \n",
    "                info = line.strip().split('\\t')\n",
    "                \n",
    "                chrom = info[0]\n",
    "                start = int(info[1])\n",
    "                end = int(info[2])\n",
    "                seq = str(info[3])\n",
    "                count = float(info[7])\n",
    "                strand = info[5]\n",
    "\n",
    "                if len(seq) >= min_len and len(seq) <= max_len : \n",
    "                    if seq[0] in p5nt : \n",
    "                        if strand == orientation : \n",
    "                            lines += f\"{chrom}\\t{start}\\t{end}\\t{seq}\\t{count}\\t{strand}\\n\"\n",
    "    f.close() \n",
    "    \n",
    "    output.write(lines)\n",
    "    \n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6b8ae9c5-75d6-4e98-a72b-c0286c54273d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def intersect_bed(bed1, bed2, output) : \n",
    "    \n",
    "    os.system(f\"sh intersect.sh {bed1} {bed2} {output}\")\n",
    "    \n",
    "    return output\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "38ea1339-963b-4313-bc1d-3d9369c9e982",
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_total_lines(file) : \n",
    "    \n",
    "    i = 0\n",
    "    with open(file, 'r') as f : \n",
    "        for line in f : \n",
    "            i += 1\n",
    "    f.close() \n",
    "    \n",
    "    return i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "d723ac22-7c29-4dc4-8596-39d5ebc359f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def group_and_agg_dist(df, dist_col) : \n",
    "    \n",
    "    grped = df.groupby(\n",
    "        [dist_col]\n",
    "    ).agg(\n",
    "        sense_rpm = ('sense_rpm', 'sum'), \n",
    "        anti_rpm = ('anti_rpm', 'sum')\n",
    "    ).reset_index()\n",
    "    \n",
    "    \n",
    "    grped['sense_zscore'] = 2**((grped['sense_rpm'] - grped['sense_rpm'].mean())/grped['sense_rpm'].std(ddof=0))\n",
    "    \n",
    "    grped['anti_zscore'] = 2**((grped['anti_rpm'] - grped['anti_rpm'].mean())/grped['anti_rpm'].std(ddof=0))\n",
    "    \n",
    "    grped = grped.rename(columns = {f'{dist_col}' : 'dist'})\n",
    "    \n",
    "    grped['id'] = dist_col\n",
    "    \n",
    "    return grped\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2b0f8e33-ff37-4d70-8943-31b8a753e814",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_p_distance(intersect, targets = None) :\n",
    "    \n",
    "    # calculate \n",
    "    # 5p to 5p anchored at 22G 5' end\n",
    "    # 5p to 3p\n",
    "    dat = pd.read_csv(intersect,\n",
    "                     usecols = [0, 1, 2, 4, 7, 8, 10, 12],\n",
    "                     names = ['gene', 'sense_5p', 'sense_3p', 'sense_rpm', 'anti_3p', 'anti_5p', 'anti_rpm', 'overlap'],\n",
    "                     header = None,\n",
    "                     sep = \"\\t\"\n",
    "                    )\n",
    "    \n",
    "    if targets : \n",
    "        df_filt = dat.query('gene in @targets').reset_index()\n",
    "    else : \n",
    "        df_filt = dat\n",
    "    \n",
    "    df_filt['5pto5p'] = df_filt['anti_5p'] - df_filt['sense_5p']\n",
    "    \n",
    "    df_filt['5pto3p'] = df_filt['anti_5p'] - df_filt['sense_3p']\n",
    "    \n",
    "    grouped_5p5p = group_and_agg_dist(df_filt, '5pto5p')\n",
    "    \n",
    "    grouped_5p3p = group_and_agg_dist(df_filt, '5pto3p')\n",
    "    \n",
    "    grouped = pd.concat([grouped_5p5p, grouped_5p3p])\n",
    "    \n",
    "    return grouped\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "64c21db9-993c-4774-b3cf-afad19205893",
   "metadata": {},
   "outputs": [],
   "source": [
    "def measure_dist_adj_reads(bedfile, targets = None) : \n",
    "    \n",
    "    df = pd.read_csv(bedfile, names = ['gene', 'start', 'end', 'seq', 'rpm', 'strand'], sep = \"\\t\", header = None)\n",
    "    \n",
    "    df = df.sort_values(by = ['end'], ascending = False).reset_index(drop = True)\n",
    "    \n",
    "    df['seq_len'] = df['seq'].str.len()\n",
    "    \n",
    "    df = df.query('seq_len == 22').reset_index(drop = True)\n",
    "    \n",
    "    if targets : \n",
    "        df = df.query('gene in @targets').reset_index(drop = True) \n",
    "        \n",
    "    genes = df.gene.unique().tolist()\n",
    "    tot_gene = len(genes)\n",
    "    \n",
    "    for i,gene in enumerate(genes) : \n",
    "        \n",
    "        sub = df.query('gene == @gene').reset_index(drop = True)\n",
    "\n",
    "        sub['3p'] = sub['start']\n",
    "\n",
    "        sub['5p_next'] = sub['end'].shift(-1)\n",
    "\n",
    "        sub['3p_5p'] = sub['3p'] - sub['5p_next']\n",
    "\n",
    "        sub = sub.query('`3p_5p` >= -25 and `3p_5p` <= 25')\n",
    "\n",
    "        if i == 0 : \n",
    "            res = sub\n",
    "        else : \n",
    "            res = pd.concat([res, sub], ignore_index = True)\n",
    "\n",
    "        print(f'{i+1} of {tot_gene} ({ round(( (i+1)/(tot_gene) )*100,2) }%) total genes', end='\\r')\n",
    "\n",
    "    res_group = res.groupby(\n",
    "        ['3p_5p']\n",
    "    ).agg(\n",
    "        count = ('rpm', 'sum')\n",
    "    ).reset_index() \n",
    "    \n",
    "    res_group['zscore'] = 2**((res_group['count'] - res_group['count'].mean())/res_group['count'].std(ddof=0))\n",
    "    \n",
    "    return res_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "b17b90f7-1162-4b4c-900f-8ec26a5ecf77",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_phasing(bedfile, targets = None) :\n",
    "    \n",
    "    outfile = './tmp/phasing_intersect.bed'\n",
    "    os.system(f\"sh ./phasing.sh {bedfile} {outfile}\")\n",
    "    \n",
    "    dat = pd.read_csv(outfile,\n",
    "                     usecols = [0, 1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12, 13],\n",
    "                     names = ['gene', 'up_3p', 'up_5p', 'up_seq', 'up_rpm', 'up_strand',\n",
    "                              'down_3p', 'down_5p', 'down_seq', 'down_rpm', 'down_strand',\n",
    "                              'overlap', 'p3p5_distance'],\n",
    "                     header = None,\n",
    "                     sep = \"\\t\"\n",
    "                    )\n",
    "    print(dat.head())\n",
    "    dat['rpm'] = dat['up_rpm'] + dat['down_rpm']\n",
    "    \n",
    "    if targets : \n",
    "        dat = dat.query('gene in @targets').reset_index(drop = True)\n",
    "    \n",
    "    dat_grouped = dat.groupby(\n",
    "        ['p3p5_distance']\n",
    "    ).agg(\n",
    "        count = ('rpm', 'sum')\n",
    "    ).reset_index()\n",
    "                    \n",
    "    dat_grouped['zscore'] = 2**((dat_grouped['count'] - dat_grouped['count'].mean())/dat_grouped['count'].std(ddof=0))\n",
    "    \n",
    "    return dat_grouped\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "df86aa26-70f5-4b42-b6c7-2cc5c811404f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run(bed, targets = None, output = \"./results\",  d = False) :\n",
    "    \n",
    "    mt1 = time.time()\n",
    "    \n",
    "    if not os.path.exists(output) : \n",
    "        os.mkdir(output)\n",
    "        \n",
    "    if not os.path.exists(\"./tmp\") : \n",
    "        os.mkdir(\"./tmp\")\n",
    "    \n",
    "    if d : \n",
    "        BEDFILES = glob.glob(f\"{bed}/*csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_IPArgonaute*.bed.tsv\")\n",
    "    else :\n",
    "        BEDFILES = [ bed ] \n",
    "\n",
    "    print(BEDFILES)\n",
    "\n",
    "    runtime = []\n",
    "    #BEDFILES = BEDFILES[0:2] + [ i for i in BEDFILES if \"csr\" in i ]\n",
    "    for i,B in enumerate(BEDFILES) : \n",
    "        \n",
    "        t1 = time.time() \n",
    "        \n",
    "        print(f\"Processing {os.path.basename(B)}...\\n\")\n",
    "        \n",
    "        sen = './tmp/sense.bed'\n",
    "        ant = './tmp/anti.bed'\n",
    "        inter = './tmp/inter.bed'\n",
    "        \n",
    "        #print(\"Filtering sense reads...\")\n",
    "        #filter_bed(B, sen, p5nt = ['A', 'G', 'C', 'T'], min_len = 15, max_len = 25, orientation = '+')\n",
    "        \n",
    "        #print(\"Filtering antisense reads...\")\n",
    "        filter_bed(B, ant, p5nt = ['A', 'G'], min_len = 21, max_len = 23, orientation = '-')\n",
    "        \n",
    "        #print(\"Intersecting sense to antisense reads...\")\n",
    "        #intersect = intersect_bed(sen, ant, inter)\n",
    "        \n",
    "        #anti_to_sense = calc_p_distance(intersect, targets)\n",
    "        \n",
    "        print(\"Measuring 5' to 3' distance of adjacent antisense reads...\")\n",
    "        phasing = calc_phasing(ant, targets)\n",
    "        \n",
    "        phasing['id'] = '3pto5p_anti_phasing'\n",
    "        #phasing['sense_rpm'] = 0\n",
    "        #phasing['sense_zscore'] = 0\n",
    "        restmp = phasing.rename(columns = {'count' : 'anti_rpm', 'zscore' : 'anti_zscore', 'p3p5_distance':'dist'})\n",
    "        \n",
    "        #restmp = pd.concat([anti_to_sense, phasing])\n",
    "        \n",
    "        restmp['sample'] = os.path.basename(B).replace(\".bed.tsv\", \"\")\n",
    "        \n",
    "        if i == 0 : \n",
    "            res = restmp\n",
    "        else : \n",
    "            res = pd.concat([res, restmp])\n",
    "        \n",
    "        t2 = time.time() \n",
    "        \n",
    "        runtime.append(round(t2-t1,3))\n",
    "        #if i == 3 : \n",
    "        #    print(f\"Estimated time to run pipeline for {len(BEDFILES)} bedfiles is { round( round( sum(runtime) / len(runtime), 2 ) * len(BEDFILES), 2) / 60 }\")\n",
    "        \n",
    "        print(f\"\\nProcessing {os.path.basename(B)} took {round(t2 - t1,3)} s.\\n\")\n",
    "        \n",
    "    mt2 = time.time()\n",
    "    print(f\"\\nPipeline took {round(mt2 - mt1,3)} s to complete.\\n\")\n",
    "    return res\n",
    "        \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "10cae4ec-9031-443f-be2a-b2643efe6050",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_phasing(df) : \n",
    "    \n",
    "    p = ( \n",
    "        ggplot(df) + \n",
    "            geom_bar(aes(x='3p_5p', y = 'zscore'), stat = 'identity', color = 'white') + \n",
    "            theme_classic() + \n",
    "            theme(figure_size=(8,8), \n",
    "                 aspect_ratio = 1) + \n",
    "            scale_x_continuous(limits = (-25,25), breaks = np.arange(-25,25,5))\n",
    "            \n",
    "        )\n",
    "    \n",
    "    return p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1666f896-c5f1-4cd6-a788-0331dbc34d7c",
   "metadata": {},
   "source": [
    "# Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e9a874b-b7a1-4a3f-a40a-94f9808694a8",
   "metadata": {},
   "source": [
    "### Read in targets file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "6bb8a9e9-9be6-4352-b5d9-e9cd9e6f1e47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene\tstart\tend\tseq\tcount\tstrand\tfeature\tcount_total_norm\n",
      "CELE45\t50\t72\tGTGAGGCCGGACTTGAACTCGT\t1.0\t-\tanti_transcriptome\t0.03940094177709545\n",
      "CELE45\t51\t73\tGGTGAGGCCGGACTTGAACTCG\t3.0\t-\tanti_transcriptome\t0.11820282533128636\n",
      "CELE45\t53\t74\tGGGTGAGGCCGGACTTGAACT\t1.0\t-\tanti_transcriptome\t0.03940094177709545\n",
      "CELE45\t53\t75\tGGGGTGAGGCCGGACTTGAACT\t1.0\t-\tanti_transcriptome\t0.03940094177709545\n",
      "CELE45\t53\t76\tGGGGGTGAGGCCGGACTTGAACT\t1.0\t-\tanti_transcriptome\t0.03940094177709545\n",
      "CELE45\t54\t75\tGGGGTGAGGCCGGACTTGAAC\t1.0\t-\tanti_transcriptome\t0.03940094177709545\n",
      "CELE45\t56\t77\tAGGGGGTGAGGCCGGACTTGA\t1.0\t-\tanti_transcriptome\t0.03940094177709545\n",
      "CELE45\t62\t83\tGAACCTAGGGGGTGAGGCCGG\t1.0\t-\tanti_transcriptome\t0.03940094177709545\n",
      "CELE45\t71\t93\tGAGGCTGGGTGAACCTAGGGGG\t13.0\t-\tanti_transcriptome\t0.5122122431022409\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "head \"/fs/ess/PCON0160/ben/projects/2025_csr1_cleavage/00AA_AGO_cleavage/20250421/transcripts/bed/tor123GFP::3xFLAG::sago1_V_IPArgonaute_2.trimmed.uniq.xartifacts.v2.m10.transcripts.bed.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "90a8e9fb-33a4-4a3b-9f2e-bdeed8eb7cd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "targets = '/fs/ess/PCON0160/ben/projects/2023_claycomb_argonomics/target_lists.txt'\n",
    "\n",
    "target_df = pd.read_csv(targets, sep = \"\\t\", names = ['ago', 'target'], header = None)\n",
    "target_df\n",
    "\n",
    "my_targets = target_df.query('ago == \"CSR-1\"').reset_index(drop = True).target.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f35248f8-60e5-4638-ae74-66ebbdc52609",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed = \"/fs/ess/PCON0160/ben/projects/2025_csr1_cleavage/00AA_AGO_cleavage/20250421/transcripts/bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "2b11b0d2-4b64-4539-a480-4a8baea63dd9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/fs/ess/PCON0160/ben/projects/2025_csr1_cleavage/00AA_AGO_cleavage/20250421/transcripts/bed/csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_IPArgonaute_1.trimmed.uniq.xartifacts.v2.m10.transcripts.bed.tsv', '/fs/ess/PCON0160/ben/projects/2025_csr1_cleavage/00AA_AGO_cleavage/20250421/transcripts/bed/csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_IPArgonaute_2.trimmed.uniq.xartifacts.v2.m10.transcripts.bed.tsv']\n",
      "Processing csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_IPArgonaute_1.trimmed.uniq.xartifacts.v2.m10.transcripts.bed.tsv...\n",
      "\n",
      "Measuring 5' to 3' distance of adjacent antisense reads...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Lmod is automatically replacing \"python/3.12\" with \"miniconda3/24.1.2-py310\".\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     gene  up_3p  up_5p                  up_seq    up_rpm up_strand  down_3p  \\\n",
      "0  CELE45     61     93  GAGGCTGGGTGAACCTAGGGGG  0.087221         -       63   \n",
      "1  CELE45     61     93  GAGGCTGGGTGAACCTAGGGGG  0.087221         -       61   \n",
      "2  CELE45     61     93  GAGGCTGGGTGAACCTAGGGGG  0.087221         -       62   \n",
      "3  CELE45     61     93  GAGGCTGGGTGAACCTAGGGGG  0.087221         -       53   \n",
      "4  CELE45     61     93  GAGGCTGGGTGAACCTAGGGGG  0.087221         -       54   \n",
      "\n",
      "   down_5p                 down_seq  down_rpm down_strand  overlap  \\\n",
      "0       85   GTGAACCTAGGGGGTGAGGCCG  0.087221           -       22   \n",
      "1       83   GAACCTAGGGGGTGAGGCCGGA  0.959432           -       22   \n",
      "2       83    GAACCTAGGGGGTGAGGCCGG  0.087221           -       21   \n",
      "3       76  GGGGGTGAGGCCGGACTTGAACT  0.174442           -       15   \n",
      "4       76   GGGGGTGAGGCCGGACTTGAAC  0.087221           -       15   \n",
      "\n",
      "   p3p5_distance  \n",
      "0            -14  \n",
      "1            -12  \n",
      "2            -12  \n",
      "3             -5  \n",
      "4             -5  \n",
      "\n",
      "Processing csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_IPArgonaute_1.trimmed.uniq.xartifacts.v2.m10.transcripts.bed.tsv took 41.326 s.\n",
      "\n",
      "Processing csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_IPArgonaute_2.trimmed.uniq.xartifacts.v2.m10.transcripts.bed.tsv...\n",
      "\n",
      "Measuring 5' to 3' distance of adjacent antisense reads...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Lmod is automatically replacing \"python/3.12\" with \"miniconda3/24.1.2-py310\".\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     gene  up_3p  up_5p                   up_seq    up_rpm up_strand  down_3p  \\\n",
      "0  CELE45     60     93  GAGGCTGGGTGAACCTAGGGGGT  0.093513         -       62   \n",
      "1  CELE45     60     93  GAGGCTGGGTGAACCTAGGGGGT  0.093513         -       63   \n",
      "2  CELE45     60     93  GAGGCTGGGTGAACCTAGGGGGT  0.093513         -       61   \n",
      "3  CELE45     60     93  GAGGCTGGGTGAACCTAGGGGGT  0.093513         -       62   \n",
      "4  CELE45     60     93  GAGGCTGGGTGAACCTAGGGGGT  0.093513         -       55   \n",
      "\n",
      "   down_5p                 down_seq  down_rpm down_strand  overlap  \\\n",
      "0       85  GTGAACCTAGGGGGTGAGGCCGG  0.046756           -       23   \n",
      "1       85   GTGAACCTAGGGGGTGAGGCCG  0.093513           -       22   \n",
      "2       83   GAACCTAGGGGGTGAGGCCGGA  0.420807           -       22   \n",
      "3       83    GAACCTAGGGGGTGAGGCCGG  0.093513           -       21   \n",
      "4       77   AGGGGGTGAGGCCGGACTTGAA  0.046756           -       17   \n",
      "\n",
      "   p3p5_distance  \n",
      "0            -15  \n",
      "1            -15  \n",
      "2            -13  \n",
      "3            -13  \n",
      "4             -7  \n",
      "\n",
      "Processing csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_IPArgonaute_2.trimmed.uniq.xartifacts.v2.m10.transcripts.bed.tsv took 74.403 s.\n",
      "\n",
      "\n",
      "Pipeline took 115.729 s to complete.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "### me_result = run(bed = bed, d = True, targets = my_targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "53ed2944-45ca-4764-aab5-e1f16bbaf9a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "me_result.to_csv(\"/fs/ess/PCON0160/ben/projects/2025_csr1_cleavage/phasing_analysis.tsv\", sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "a7f74953-815a-4526-87cd-2748119f0951",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dist</th>\n",
       "      <th>anti_rpm</th>\n",
       "      <th>anti_zscore</th>\n",
       "      <th>id</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-21</td>\n",
       "      <td>0.523326</td>\n",
       "      <td>0.613454</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "      <td>csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-20</td>\n",
       "      <td>19.014194</td>\n",
       "      <td>2.404691</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "      <td>csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-19</td>\n",
       "      <td>37.766725</td>\n",
       "      <td>9.610186</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "      <td>csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-18</td>\n",
       "      <td>23.811353</td>\n",
       "      <td>3.427484</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "      <td>csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-17</td>\n",
       "      <td>19.363079</td>\n",
       "      <td>2.467477</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "      <td>csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>5</td>\n",
       "      <td>0.794857</td>\n",
       "      <td>0.559840</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "      <td>csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>6</td>\n",
       "      <td>0.935126</td>\n",
       "      <td>0.566309</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "      <td>csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>7</td>\n",
       "      <td>1.917009</td>\n",
       "      <td>0.613732</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "      <td>csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>8</td>\n",
       "      <td>0.794857</td>\n",
       "      <td>0.559840</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "      <td>csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>9</td>\n",
       "      <td>2.571597</td>\n",
       "      <td>0.647534</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "      <td>csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    dist   anti_rpm  anti_zscore                   id  \\\n",
       "0    -21   0.523326     0.613454  3pto5p_anti_phasing   \n",
       "1    -20  19.014194     2.404691  3pto5p_anti_phasing   \n",
       "2    -19  37.766725     9.610186  3pto5p_anti_phasing   \n",
       "3    -18  23.811353     3.427484  3pto5p_anti_phasing   \n",
       "4    -17  19.363079     2.467477  3pto5p_anti_phasing   \n",
       "..   ...        ...          ...                  ...   \n",
       "26     5   0.794857     0.559840  3pto5p_anti_phasing   \n",
       "27     6   0.935126     0.566309  3pto5p_anti_phasing   \n",
       "28     7   1.917009     0.613732  3pto5p_anti_phasing   \n",
       "29     8   0.794857     0.559840  3pto5p_anti_phasing   \n",
       "30     9   2.571597     0.647534  3pto5p_anti_phasing   \n",
       "\n",
       "                                               sample  \n",
       "0   csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...  \n",
       "1   csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...  \n",
       "2   csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...  \n",
       "3   csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...  \n",
       "4   csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...  \n",
       "..                                                ...  \n",
       "26  csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...  \n",
       "27  csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...  \n",
       "28  csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...  \n",
       "29  csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...  \n",
       "30  csr1tor104csr1_exon1::GFP::Flag_IV:7957568_IV_...  \n",
       "\n",
       "[61 rows x 5 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "me_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bbab8ab-a9cc-47bc-a5b2-19a32d06da98",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.12 (Conda 2024.06) [python/3.12]",
   "language": "python",
   "name": "python312"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
