{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "e46a429c-b4dd-416f-bc87-98874700693d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd \n",
    "import argparse\n",
    "import numpy as np \n",
    "import glob\n",
    "import time \n",
    "import logging\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle\n",
    "from plotnine import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "0b8ee29d-8559-4779-9c3e-750816518ffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_bed(bed, outname, p5nt = ['G', 'A'], min_len = 21, max_len = 23, orientation = '-') : \n",
    "    \n",
    "    output = open(outname, 'w')\n",
    "    lines = ''\n",
    "    with open(bed, 'r') as f : \n",
    "        for line in f :\n",
    "            if not line.startswith(\"gene_\") : \n",
    "                info = line.strip().split('\\t')\n",
    "\n",
    "                chrom = info[0]\n",
    "                start = int(info[1])\n",
    "                end = int(info[2])\n",
    "                seq = str(info[3])\n",
    "                count = float(info[7])\n",
    "                strand = info[5]\n",
    "\n",
    "                if len(seq) >= min_len and len(seq) <= max_len : \n",
    "                    if seq[0] in p5nt : \n",
    "                        if strand == orientation : \n",
    "                            lines += f\"{chrom}\\t{start}\\t{end}\\t{seq}\\t{count}\\t{strand}\\n\"\n",
    "    f.close() \n",
    "    \n",
    "    output.write(lines)\n",
    "    \n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "6b8ae9c5-75d6-4e98-a72b-c0286c54273d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def intersect_bed(bed1, bed2, output) : \n",
    "    \n",
    "    os.system(f\"sh intersect.sh {bed1} {bed2} {output}\")\n",
    "    \n",
    "    return output\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "38ea1339-963b-4313-bc1d-3d9369c9e982",
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_total_lines(file) : \n",
    "    \n",
    "    i = 0\n",
    "    with open(file, 'r') as f : \n",
    "        for line in f : \n",
    "            i += 1\n",
    "    f.close() \n",
    "    \n",
    "    return i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "d723ac22-7c29-4dc4-8596-39d5ebc359f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def group_and_agg_dist(df, dist_col) : \n",
    "    \n",
    "    grped = df.groupby(\n",
    "        [dist_col]\n",
    "    ).agg(\n",
    "        sense_rpm = ('sense_rpm', 'sum'), \n",
    "        anti_rpm = ('anti_rpm', 'sum')\n",
    "    ).reset_index()\n",
    "    \n",
    "    \n",
    "    grped['sense_zscore'] = 2**((grped['sense_rpm'] - grped['sense_rpm'].mean())/grped['sense_rpm'].std(ddof=0))\n",
    "    \n",
    "    grped['anti_zscore'] = 2**((grped['anti_rpm'] - grped['anti_rpm'].mean())/grped['anti_rpm'].std(ddof=0))\n",
    "    \n",
    "    grped = grped.rename(columns = {f'{dist_col}' : 'dist'})\n",
    "    \n",
    "    grped['id'] = dist_col\n",
    "    \n",
    "    return grped\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "2b0f8e33-ff37-4d70-8943-31b8a753e814",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_p_distance(intersect, targets = None) :\n",
    "    \n",
    "    # calculate \n",
    "    # 5p to 5p anchored at 22G 5' end\n",
    "    # 5p to 3p\n",
    "    dat = pd.read_csv(intersect,\n",
    "                     usecols = [0, 1, 2, 4, 7, 8, 10, 12],\n",
    "                     names = ['gene', 'sense_5p', 'sense_3p', 'sense_rpm', 'anti_3p', 'anti_5p', 'anti_rpm', 'overlap'],\n",
    "                     header = None,\n",
    "                     sep = \"\\t\"\n",
    "                    )\n",
    "    \n",
    "    if targets : \n",
    "        df_filt = dat.query('gene in @targets').reset_index()\n",
    "    else : \n",
    "        df_filt = dat\n",
    "    \n",
    "    df_filt['5pto5p'] = df_filt['anti_5p'] - df_filt['sense_5p']\n",
    "    \n",
    "    df_filt['5pto3p'] = df_filt['anti_5p'] - df_filt['sense_3p']\n",
    "    \n",
    "    grouped_5p5p = group_and_agg_dist(df_filt, '5pto5p')\n",
    "    \n",
    "    grouped_5p3p = group_and_agg_dist(df_filt, '5pto3p')\n",
    "    \n",
    "    grouped = pd.concat([grouped_5p5p, grouped_5p3p])\n",
    "    \n",
    "    return grouped\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "64c21db9-993c-4774-b3cf-afad19205893",
   "metadata": {},
   "outputs": [],
   "source": [
    "def measure_dist_adj_reads(bedfile, targets = None) : \n",
    "    \n",
    "    df = pd.read_csv(bedfile, names = ['gene', 'start', 'end', 'seq', 'rpm', 'strand'], sep = \"\\t\", header = None)\n",
    "    \n",
    "    df['seq_len'] = df['seq'].str.len()\n",
    "        \n",
    "    df = df.query('seq_len == 22').reset_index(drop = True)\n",
    "    \n",
    "    if targets : \n",
    "        df = df.query('gene in @targets').reset_index(drop = True) \n",
    "        \n",
    "    genes = df.gene.unique().tolist()\n",
    "    tot_gene = len(genes)\n",
    "    \n",
    "    for i,gene in enumerate(genes) : \n",
    "        \n",
    "        sub = df.query('gene == @gene').sort_values(by = ['end'], ascending = False)\n",
    "\n",
    "        sub['3p'] = sub['start']\n",
    "\n",
    "        sub['5p_next'] = sub['end'].shift(-1)\n",
    "\n",
    "\n",
    "        sub['3p_5p'] = sub['3p'] - sub['5p_next']\n",
    "\n",
    "        sub = sub.query('`3p_5p` >= -25 and `3p_5p` <= 25')\n",
    "\n",
    "        if i == 0 : \n",
    "            res = sub\n",
    "        else : \n",
    "            res = pd.concat([res, sub], ignore_index = True)\n",
    "\n",
    "\n",
    "        print(f'{i+1} of {tot_gene} ({ round(( (i+1)/(tot_gene) )*100,2) }%) total genes', end='\\r')\n",
    "\n",
    "    res_group = res.groupby(\n",
    "        ['3p_5p']\n",
    "    ).agg(\n",
    "        count = ('rpm', 'sum')\n",
    "    ).reset_index() \n",
    "    \n",
    "    res_group['zscore'] = 2**((res_group['count'] - res_group['count'].mean())/res_group['count'].std(ddof=0))\n",
    "    \n",
    "    return res_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "df86aa26-70f5-4b42-b6c7-2cc5c811404f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run(bed, targets = None, output = \"./results\",  d = False) :\n",
    "    \n",
    "    mt1 = time.time()\n",
    "    \n",
    "    if not os.path.exists(output) : \n",
    "        os.mkdir(output)\n",
    "        \n",
    "    if not os.path.exists(\"./tmp\") : \n",
    "        os.mkdir(\"./tmp\")\n",
    "    \n",
    "    if d : \n",
    "        BEDFILES = glob.glob(f\"{bed}/*.bed.tsv\")\n",
    "    else :\n",
    "        BEDFILES = [ bed ] \n",
    "\n",
    "    for i,B in enumerate(BEDFILES) : \n",
    "        \n",
    "        t1 = time.time() \n",
    "        \n",
    "        print(f\"Processing {os.path.basename(B)}...\\n\")\n",
    "        \n",
    "        sen = './tmp/sense.bed'\n",
    "        ant = './tmp/anti.bed'\n",
    "        inter = './tmp/inter.bed'\n",
    "        \n",
    "        print(\"Filtering sense reads...\")\n",
    "        filter_bed(B, sen, p5nt = ['A', 'G', 'C', 'T'], min_len = 15, max_len = 25, orientation = '+')\n",
    "        \n",
    "        print(\"Filtering antisense reads...\")\n",
    "        filter_bed(B, ant, p5nt = ['A', 'G'], min_len = 21, max_len = 23, orientation = '-')\n",
    "        \n",
    "        print(\"Intersecting sense to antisense reads...\")\n",
    "        intersect = intersect_bed(sen, ant, inter)\n",
    "        \n",
    "        anti_to_sense = calc_p_distance(intersect, targets)\n",
    "        \n",
    "        print(\"Measuring 5' to 3' distance of adjacent antisense reads...\")\n",
    "        phasing = measure_dist_adj_reads(ant, targets)\n",
    "        \n",
    "        phasing['id'] = '3pto5p_anti_phasing'\n",
    "        phasing['sense_rpm'] = 0\n",
    "        phasing['sense_zscore'] = 0\n",
    "        phasing = phasing.rename(columns = {'count' : 'anti_rpm', 'zscore' : 'anti_zscore', '3p_5p':'dist'})\n",
    "        \n",
    "        restmp = pd.concat([anti_to_sense, phasing])\n",
    "        \n",
    "        if i == 0 : \n",
    "            res = restmp\n",
    "        else : \n",
    "            res = pd.concat([res, restmp])\n",
    "        \n",
    "        t2 = time.time() \n",
    "        print(f\"\\nProcessing {os.path.basename(B)} took {round(t2 - t1,3)} s.\\n\")\n",
    "        \n",
    "    mt2 = time.time()\n",
    "    print(f\"\\nPipeline took {round(mt2 - mt1,3)} s to complete.\\n\")\n",
    "    return res\n",
    "        \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "10cae4ec-9031-443f-be2a-b2643efe6050",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_phasing(df) : \n",
    "    \n",
    "    p = ( \n",
    "        ggplot(df) + \n",
    "            geom_bar(aes(x='3p_5p', y = 'zscore'), stat = 'identity', color = 'white') + \n",
    "            theme_classic() + \n",
    "            theme(figure_size=(8,8), \n",
    "                 aspect_ratio = 1) + \n",
    "            scale_x_continuous(limits = (-25,25), breaks = np.arange(-25,25,5))\n",
    "            \n",
    "        )\n",
    "    \n",
    "    return p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1666f896-c5f1-4cd6-a788-0331dbc34d7c",
   "metadata": {},
   "source": [
    "# Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e9a874b-b7a1-4a3f-a40a-94f9808694a8",
   "metadata": {},
   "source": [
    "### Read in targets file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "90a8e9fb-33a4-4a3b-9f2e-bdeed8eb7cd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['WBGene00022277',\n",
       " 'WBGene00022278',\n",
       " 'WBGene00022279',\n",
       " 'WBGene00021677',\n",
       " 'WBGene00000812',\n",
       " 'WBGene00021682',\n",
       " 'WBGene00021676',\n",
       " 'WBGene00021681',\n",
       " 'WBGene00004274',\n",
       " 'WBGene00004418']"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets = \"/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/csr1_targets.txt\"\n",
    "\n",
    "csr_targets = pd.read_csv(targets, sep = \"\\t\", names = ['gene'], header = None)\n",
    "\n",
    "csr_targets_list = csr_targets.gene.tolist()\n",
    "\n",
    "csr_targets_list[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "f35248f8-60e5-4638-ae74-66ebbdc52609",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed = '/fs/ess/PAS1473/znfx1_CSRIP_WAGO9IP/metagene_alignment_03242023/transcripts/csr1_IP_2.trimmed.uniq.xc.v0.m1000.transcripts.bed.tsv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "2b11b0d2-4b64-4539-a480-4a8baea63dd9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing csr1_IP_2.trimmed.uniq.xc.v0.m1000.transcripts.bed.tsv...\n",
      "\n",
      "Filtering sense reads...\n",
      "Filtering antisense reads...\n",
      "Intersecting sense to antisense reads...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Measuring 5' to 3' distance of adjacent antisense reads...\n",
      "4007 of 4007 (100.0%) total genes\n",
      "Processing csr1_IP_2.trimmed.uniq.xc.v0.m1000.transcripts.bed.tsv took 121.3 s.\n",
      "\n",
      "\n",
      "Pipeline took 121.301 s to complete.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dist</th>\n",
       "      <th>sense_rpm</th>\n",
       "      <th>anti_rpm</th>\n",
       "      <th>sense_zscore</th>\n",
       "      <th>anti_zscore</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1279.390528</td>\n",
       "      <td>6160.910293</td>\n",
       "      <td>0.765386</td>\n",
       "      <td>0.709430</td>\n",
       "      <td>5pto5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1677.405916</td>\n",
       "      <td>11697.295965</td>\n",
       "      <td>1.365065</td>\n",
       "      <td>1.893398</td>\n",
       "      <td>5pto5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1788.482484</td>\n",
       "      <td>8672.593164</td>\n",
       "      <td>1.604271</td>\n",
       "      <td>1.107447</td>\n",
       "      <td>5pto5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>1776.534333</td>\n",
       "      <td>7744.577005</td>\n",
       "      <td>1.576647</td>\n",
       "      <td>0.939422</td>\n",
       "      <td>5pto5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>1582.179523</td>\n",
       "      <td>7213.375533</td>\n",
       "      <td>1.188600</td>\n",
       "      <td>0.854978</td>\n",
       "      <td>5pto5p</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>21.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>230.256887</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.655638</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>22.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>214.562378</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.653357</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>23.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>192.564295</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.650173</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>24.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>207.530005</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.652337</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>25.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>196.896947</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.650799</td>\n",
       "      <td>3pto5p_anti_phasing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>141 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    dist    sense_rpm      anti_rpm  sense_zscore  anti_zscore  \\\n",
       "0    1.0  1279.390528   6160.910293      0.765386     0.709430   \n",
       "1    2.0  1677.405916  11697.295965      1.365065     1.893398   \n",
       "2    3.0  1788.482484   8672.593164      1.604271     1.107447   \n",
       "3    4.0  1776.534333   7744.577005      1.576647     0.939422   \n",
       "4    5.0  1582.179523   7213.375533      1.188600     0.854978   \n",
       "..   ...          ...           ...           ...          ...   \n",
       "42  21.0     0.000000    230.256887      0.000000     0.655638   \n",
       "43  22.0     0.000000    214.562378      0.000000     0.653357   \n",
       "44  23.0     0.000000    192.564295      0.000000     0.650173   \n",
       "45  24.0     0.000000    207.530005      0.000000     0.652337   \n",
       "46  25.0     0.000000    196.896947      0.000000     0.650799   \n",
       "\n",
       "                     id  \n",
       "0                5pto5p  \n",
       "1                5pto5p  \n",
       "2                5pto5p  \n",
       "3                5pto5p  \n",
       "4                5pto5p  \n",
       "..                  ...  \n",
       "42  3pto5p_anti_phasing  \n",
       "43  3pto5p_anti_phasing  \n",
       "44  3pto5p_anti_phasing  \n",
       "45  3pto5p_anti_phasing  \n",
       "46  3pto5p_anti_phasing  \n",
       "\n",
       "[141 rows x 6 columns]"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "me_result = run(bed = bed, targets = csr_targets_list, d = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53ed2944-45ca-4764-aab5-e1f16bbaf9a8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ebe511c-4a61-4d01-87a5-7634ba9037ad",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9 (Conda 2022.05) [python/3.9-2022.05]",
   "language": "python",
   "name": "python39_202205"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
